Add support for Tape test framework

jcansdale opened the pull request:
Issue #986

##### Fix

This pull adds support for the Tape JavaScript test framework.

There are unit tests for this in a separate project:
https://github.com/jcansdale/TestFrameworksNTVS

It would be nice to integrate these at some point (they are in a .njsproj project). I don't know where it would make sense to put them? The test framework support is useful on its own though, so I've submitted this pull request without the tests.

msftclas commented:
Hi __@jcansdale__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, MSBOT;

jcansdale commented:
Oops, I forgot to add the tape.js test framework to the installer. I'll update my branch and report back.

jcansdale commented:
Added 'TestFrameworks\Tape' folder to Wix installer file. Is this enough to include the contained 'tape.js' file? Please check this because I haven't been able to build the installer locally!

msftclas commented:
Hi __@jcansdale__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, MSBOT;

mjbvz commented on Nodejs/Product/Nodejs/TestFrameworks/Tape/tape.js:
```diff
@@ -0,0 +1,75 @@
+var fs = require('fs');
```
[nit] I'm not sure our other test frameworks do this, but please add `"use strict";` for this new file.

mjbvz commented on Nodejs/Product/Nodejs/TestFrameworks/Tape/tape.js:
```diff
@@ -0,0 +1,75 @@
+    testFileList.split(';').forEach(function (testFile) {
+
+        var testCases;
+        process.chdir(path.dirname(testFile));
```
Does `require(testFile)` without changing the directory not work here?

mjbvz replied:
Sorry, misinterpreted the purpose of changing the working directory here. This is what we do for the other test framework adapters so this looks correct.

jcansdale replied:
I did the same as the mocha.js framework here. It is related to Fix #8 - "test explorer doesn't run mocha tests with fs.readFileSync and possibly more". This could be an issue with Tape style tests as well.

mjbvz commented on Nodejs/Product/Nodejs/TestFrameworks/Tape/tape.js:
```diff
@@ -0,0 +1,75 @@
+        } catch (ex) {
+            console.error("NTVS_ERROR:", ex, "in", testFile);
+            return;
+        }
```
[nit] Since this logic is duplicated, please extract it to it's own function that returns a Boolean that indicates if the test was loaded or not.

jcansdale replied:
Will do. I'll factor out the 'tape' finding logic while I'm at it.

mjbvz commented:
Thanks for looking into this! Just two minor comments and then we can merge the change in. We'll try to get this into the next NTVS 1.2 release if possible.

jcansdale commented:
Thanks for the code review! I've incorporated your suggestions.

mousetraps commented on Nodejs/Setup/NodejsTools/NodejsTools.wxs:
```diff
@@ -15,6 +15,7 @@
             <Directory Id="Dir_$(var.key)TestFrameworks" Name="TestFrameworks">
                 <Directory Id="Dir_$(var.key)TestFrameworksExportRunner" Name="ExportRunner" />
                 <Directory Id="Dir_$(var.key)TestFrameworksMocha" Name="Mocha" />
+                <Directory Id="Dir_$(var.key)TestFrameworksTape" Name="Tape" />
```
The directory is defined, but it's not being used anywhere. You'll also need to include a File Include in NodejsToolsFiles.proj
```
       <File Include="TestFrameworks\tape\tape.js">
            <InstallDirectory>TestFrameworksTape</InstallDirectory>
        </File>
```

mousetraps replied:
Also you mentioned you were unable to build the installer? What error were you running into?

jcansdale replied:
I've been able to get the installer built. I'm a PowerShell newb, but you have good documentation!

mousetraps commented on Nodejs/Product/Nodejs/TestFrameworks/Tape/tape.js:
```diff
@@ -0,0 +1,73 @@
+        var tapePath = path.join(projectFolder, 'node_modules', 'tape');
+        return require(tapePath);
+    } catch (e) {
+        console.error("NTVS_ERROR:", "Couldn't find 'tape' module relative to", projectFolder);
```
This error message isn't clear - let's make it consistent with the one in mocha.js

mousetraps commented on Nodejs/Product/Nodejs/TestFrameworks/Tape/tape.js:
```diff
@@ -0,0 +1,73 @@
+
+function find_tests(testFileList, discoverResultFile, projectFolder) {
+    var test = findTape(projectFolder);
+    if (test === null) return;
```
nit: siorry for not defining this more clearly, but just to be consistent with the rest of our codebase... if statements should use braces and no single-line ifs

mousetraps commented:
Hey @jcansdale , thanks for the contribution! We've had a lot of folks request tape support lately, so it's exciting to finally add support! Sorry for jumping into this a little late... left a few more comments. I don't think we'll be able to make the Beta ship-train, but let's definitely aim to pull this into v1.2.

jcansdale commented:
What happened to all of @mousetraps comments? I can't find them anywhere! :confused:

mjbvz commented:
Sorry, @mousetraps got flagged as a bot by github for some reason, which hides all of her comments and prs. We've contacted Github support to get this fixed as soon as possible. Everything should show up again after that.

mousetraps commented:
Had to send in some DNA, and the good news is that I've been declared human! All my comments should now be visible :smiley:

jcansdale commented:
Congratulations on your machine like speed and efficiency, but I'm glad to hear you've been confirmed as human! :wink:

jcansdale commented:
@mousetraps I've implemented your various suggestions, including:

Added Tape UnitTest file templates for JavaScript and TypeScript.
Made error messages more helpful.

If you have a chance to give it a try, I'd be interested to hear how you get on.

mousetraps commented on Nodejs/Product/Nodejs/Nodejs.csproj:
```diff
@@ -875,6 +875,10 @@
+    <ZipItem Include="Templates\Files\TapeUnitTest\UnitTest.js" />
+    <ZipItem Include="Templates\Files\TapeUnitTest\UnitTest.vstemplate" />
+    <ZipItem Include="Templates\Files\TypeScriptTapeUnitTest\UnitTest.vstemplate" />
+    <ZipItem Include="Templates\Files\TypeScriptTapeUnitTest\UnitTest.ts" />
```
One minor note which can be addressed later... we'll need to update the VS "15" template manifest to account for this change.
https://github.com/Microsoft/nodejstools/blob/master/Nodejs/Product/Nodejs/Templates/VS_Nodejs.item.vstman

mousetraps commented:
:+1:

Nice work, thanks!! We'll merge this in and start testing... hopefully it'll make 1.2 Beta! :smiley:
